test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+        --   * Every 'vcSigningKey' in all its certificates has a corresponding
+        --    'StakeholderId' in 'mrs HM.! epoch'
+        --   * The set of its 'StakeholderId' keys is a subset of all the
+        --    'StakeholderId's in 'mrs'
+        certsMap <- do
+            let vssGen :: Gen VssCertificate
+                vssGen = do
+                    -- This list is guaranteed to be non-empty
+                    richman <- elements richmenPks
+                    vssCert <- arbitrary
+                    return $ vssCert {vcSigningKey = richman}
+                gen :: StakeholderId -> Gen (StakeholderId, VssCertificate)
+                gen sid = (,) <$> pure sid <*> vssGen
+            participantsMap <- HM.fromList <$> mapM gen participants
+            fillerMap <- customHashMapGen arbitrary vssGen
+            return $ HM.union participantsMap fillerMap
+
+        -- The 'VssCertificatesMap' field of this 'VssCertData' value satisfies:
+        --   * None of its 'StakeholderId' keys is a richman
+        --    (i.e. is a  member of 'richmen')
+        _gsVssCertificates <- do
+            certs <- customHashMapGen
+                (arbitrary `suchThat` (not . flip HM.member richmen))
+                arbitrary
+            vssData <- arbitrary
+            return $ vssData {certs = certs}
+
+        return (epoch, GtGlobalState {..}, certsMap, mrs)
+
+checksGoodCertsPayload :: GoodCertsPayload -> Bool
+checksGoodCertsPayload (getGoodCerts -> (epoch, gtgs, certsMap, mrs)) =
+    isRight . tossRunner mrs gtgs $ checkCertificatesPayload epoch certsMap
+
+checksBadCertsPayload :: GoodCertsPayload -> StakeholderId -> VssCertificate -> Property
+checksBadCertsPayload (getGoodCerts -> (epoch, gtgs, certsMap, mrs)) sid cert =
+    let mrsWithMissingEpoch = HM.delete epoch mrs
+        noRichmen =
+            tossRunner mrsWithMissingEpoch gtgs $ checkCertificatesPayload epoch certsMap
+        res1 = case noRichmen of
+            Left (NoRichmen e) -> e == epoch
+            _ -> False
+
+        insCert = HM.insert sid cert
+        newCertsMap = insCert certsMap
+        newGtgs = gtgs & gsVssCertificates %~ (\vcd -> let crt = insCert $ certs vcd
+                                                       in vcd { certs = crt })
+        certAlreadySent =
+            tossRunner mrs newGtgs  $ checkCertificatesPayload epoch newCertsMap
+        res2 = case certAlreadySent of
+            Left (CertificateAlreadySent nes) -> sid `elem` nes
+            _ -> False
+
+        sid' = addressHash . vcSigningKey $ cert
+        newerCertsMap = HM.insert sid' cert certsMap
+        certNoRichmen = tossRunner mrs gtgs $ checkCertificatesPayload epoch newerCertsMap
+        res3 = case certNoRichmen of
+            Left (CertificateNotRichmen nes) -> sid' `elem` nes
+            _ -> False
+
+    in not (HM.member sid' $ mrs HM.! epoch) ==> res1 && res2 && res3
+
+-- | Utility functions for this module
```

Comment:
You wrote a haddock comment to the `tossRunner` function. But it applies not to these functions, but to all functions below (currently there are 2 functions).

So it should be a regular comment. We often use such formatting:
```
----------------------------------------------------------------------------
-- Utility functions for this module
----------------------------------------------------------------------------
```